irc/proto/src/lib.rs

```diff
@@ -49,33 +49,48 @@ pub fn command(command: &str, parameters: Vec<String>) -> Message {
 }
 
 /// Reference: https://defs.ircdocs.horse/defs/chantypes
-pub const CHANNEL_PREFIXES: [char; 4] = ['#', '&', '+', '!'];
+pub const CHANNEL_PREFIXES: &[char] = &['#', '&', '+', '!'];
+
+/// Reference: https://defs.ircdocs.horse/defs/chantypes
+///
+/// Channel types which should be used if the CHANTYPES ISUPPORT is not returned
+pub const DEFAULT_CHANNEL_PREFIXES: &[char] = &['#', '&'];
+
 /// https://modern.ircdocs.horse/#channels
 ///
 /// Channel names are strings (beginning with specified prefix characters). Apart from the requirement of
 /// the first character being a valid channel type prefix character; the only restriction on a channel name
 /// is that it may not contain any spaces (' ', 0x20), a control G / BELL ('^G', 0x07), or a comma (',', 0x2C)
 /// (which is used as a list item separator by the protocol).
-pub const CHANNEL_BLACKLIST_CHARS: [char; 3] = [',', '\u{07}', ','];
+pub const CHANNEL_BLACKLIST_CHARS: &[char] = &[',', '\u{07}', ','];
 
-pub fn is_channel(target: &str) -> bool {
-    target.starts_with(CHANNEL_PREFIXES) && !target.contains(CHANNEL_BLACKLIST_CHARS)
+pub fn is_channel(target: &str, chantypes: &[char]) -> bool {
+    target.starts_with(chantypes) && !target.contains(CHANNEL_BLACKLIST_CHARS)
 }
 
 // Reference: https://defs.ircdocs.horse/defs/chanmembers
-pub const CHANNEL_MEMBERSHIP_PREFIXES: [char; 6] = ['~', '&', '!', '@', '%', '+'];
-
-pub fn parse_channel_from_target(target: &str) -> Option<(Option<char>, String)> {
-    if target.starts_with(CHANNEL_MEMBERSHIP_PREFIXES) {
-        let channel = target.strip_prefix(CHANNEL_MEMBERSHIP_PREFIXES)?;
+pub const CHANNEL_MEMBERSHIP_PREFIXES: &[char] = &['~', '&', '!', '@', '%', '+'];
 
-        if is_channel(channel) {
-            return Some((target.chars().next(), channel.to_string()));
-        }
-    }
+/// https://modern.ircdocs.horse/#channels
+///
+/// Given a target, split it into a channel name (beginning with a character in `chantypes`) and a
+/// possible list of prefixes (given in `statusmsg_prefixes`). If these two lists overlap, the
+/// behaviour is unspecified.
+pub fn parse_channel_from_target(
+    target: &str,
+    chantypes: &[char],
+    statusmsg_prefixes: &[char],
+) -> Option<(Vec<char>, String)> {
+    // We parse the target by finding the first character in chantypes, and returing (even if that
+    // character is in statusmsg_prefixes)
+    // If the characters before the first chantypes character are all valid prefixes, then we have
+    // a valid channel name with those prefixes.    let chan_index = target.find(chantypes)?;
+    let chan_index = target.find(chantypes)?;
 
-    if is_channel(target) {
-        Some((None, target.to_string()))
+    // will not panic, since `find` always returns a valid codepoint index
+    let (prefix, chan) = target.split_at(chan_index);
```

Comment:
I changed this and my tests started failing... so now I remember why I did this :)

We want an inclusive split for this function, since the channel string includes its chantype. split_at is an inclusive split, and we don't want to split multiple times like `split_inclusive`.

Reply:
Ahh, right. There is no `split_once_inclusive`. Let's stick with the `split_at` then.